Wikipedia/Code/SchemeHandler/SchemeHandler.swift

```diff
@@ -0,0 +1,241 @@
+        var headerFields = [String: String](minimumCapacity: 1)
+        if let contentType = contentType {
+            headerFields["Content-Type"] = contentType
+        }
+        if let response = HTTPURLResponse(url: requestURL, statusCode: 200, httpVersion: nil, headerFields: headerFields) {
+            cache.cacheResponse(response, data: data, path: path)
+        }
+    }
+    
+    @objc(cacheSectionDataForArticle:)
+    func cacheSectionData(for article: MWKArticle) {
+        cache.cacheSectionData(for: article)
+    }
+}
+
+extension SchemeHandler: WKURLSchemeHandler {
+    
+    func webView(_ webView: WKWebView, start urlSchemeTask: WKURLSchemeTask) {
+        
+        assert(Thread.isMainThread)
+        
+        var request = urlSchemeTask.request
+        guard let requestURL = request.url else {
+            urlSchemeTask.didFailWithError(SchemeHandlerError.invalidParameters)
+            return
+        }
+        guard let components = NSURLComponents(url: requestURL, resolvingAgainstBaseURL: false) else {
+            urlSchemeTask.didFailWithError(SchemeHandlerError.invalidParameters)
+            return
+        }
+        #if WMF_LOCAL
+        components.scheme = components.host == "localhost" ? "http" : "https"
+        #else
+        components.scheme =  "https"
+        #endif
+        guard let pathComponents = (components.path as NSString?)?.pathComponents,
+            pathComponents.count >= 2 else {
+            urlSchemeTask.didFailWithError(SchemeHandlerError.invalidParameters)
+            return
+        }
+        
+        let baseComponent = pathComponents[1]
+        
+        let localCompletionBlock: (URLResponse?, Data?, Error?) -> Void = { (response, data, error) in
+            
+            assert(Thread.isMainThread)
+            
+            guard self.schemeTaskIsActive(urlSchemeTask: urlSchemeTask) else {
+                return
+            }
+            
+            if response == nil && error == nil {
+                urlSchemeTask.didFailWithError(SchemeHandlerError.unexpectedResponse)
+                self.activeSchemeTasks.remove(urlSchemeTask)
+                return
+            }
+            
+            if let error = error {
+                urlSchemeTask.didFailWithError(error)
+                self.activeSchemeTasks.remove(urlSchemeTask)
+                return
+            }
+            
+            if let response = response {
+                urlSchemeTask.didReceive(response)
+            }
+            
+            if let data = data {
+                urlSchemeTask.didReceive(data)
+            }
+            urlSchemeTask.didFinish()
+            self.activeSchemeTasks.remove(urlSchemeTask)
+        }
+        
+        activeSchemeTasks.add(urlSchemeTask)
+        
+        switch baseComponent {
+        case FileHandler.basePath:
+            fileHandler.handle(pathComponents: pathComponents, requestURL: requestURL, completion: localCompletionBlock)
```

Comment:
Since this reads from the filesystem, It'd be good to move it (and this entire switch statement) off of the main queue. My guess is the URLCache in the default handler might hit the filesystem as well. Callbacks to the task would have to bounce back to main.

Reply:
agh good point, will do